### PR TITLE
Rework sample sheet validation and error display

### PR DESCRIPTION
**Pull request by wow-such-code**
* Removes validation case based on duplicate conditions and replicate ids
* Removes notification popup that interfered with spreadsheet and was hard to read
* Upon "registration" button click, adds red borders to mandatory cells that were not filled
* Removes borders for cells that were filled in the meantime (also upon "registration")
* Provides means to display error messages in the Span above the spreadsheet
* Displays the instruction to fill in missing values above the spreadsheet

image

### CODE REVIEW

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java [hunk 1]
```diff
@@ -442,22 +453,48 @@ public Result<Void, InvalidSpreadsheetRow> areInputsValid() {
...
       // mandatory not filled in --> invalid
-      if (mandatoryInputs.stream().anyMatch(String::isBlank)) {
-        return Result.fromError(new InvalidSpreadsheetRow(
-            SpreadsheetInvalidationReason.MISSING_INPUT, rowId));
+      for(int colId : mandatoryInputCols) {
+        Cell cell = row.getCell(colId);
+        if(SpreadsheetMethods.cellToStringOrNull(cell).isBlank()) {
+          invalidCells.add(cell);
+        } else {
+          validCells.add(cell);
```
Comment (sven1103) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java:461]: why is it important to record the valid cells?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java [hunk 2]
```diff
@@ -442,22 +453,48 @@ public Result<Void, InvalidSpreadsheetRow> areInputsValid() {
...
+    if(!invalidCells.isEmpty()) {
+      highlightInvalidCells(invalidCells);
+
+      return Result.fromError(new InvalidSpreadsheetInput(
+          SpreadsheetInvalidationReason.MISSING_INPUT));
     }
     return Result.fromValue(null);
   }
 
+  private void highlightInvalidCells(Set<Cell> cells) {
```
Comment (sven1103) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java:478]:
```suggestion
  private void highlightInvalidCells(Collection<Cell> cells) {
```

Comment (sven1103) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java:478]: unnecessary detail

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java [hunk 3]
```diff
@@ -442,22 +453,48 @@ public Result<Void, InvalidSpreadsheetRow> areInputsValid() {
...
+    invalidStyle.setLeftBorderColor(redIndex);
+    invalidStyle.setRightBorderColor(redIndex);
+
+    for(Cell cell : cells) {
+      cell.setCellStyle(invalidStyle);
+    }
+    this.refreshCells(cells);
+  }
+
   private boolean isUniqueSampleRow(Set<String> knownIDs, Row row) {
```
Comment (sven1103) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java:498]:
```suggestion
  private boolean isUniqueSampleRow(Collection<String> knownIDs, Row row) {
```

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleSpreadsheetLayout.java [hunk 4]
```diff
@@ -151,11 +154,26 @@ private void resetSpreadSheet() {
...
-      return content.onError(error -> displayInputInvalidMessage(error.getInvalidationReason()))
+      Result<Void, InvalidSpreadsheetInput> content = sampleRegistrationSpreadsheet.areInputsValid();
+      if(content.isValue()) {
+        hideErrorInstructions();
+      }
+      return content.onError(error -> displayErrorInstructions(error.getInvalidationReason()))
           .isValue();
     }
 
+    private void displayErrorInstructions(String invalidationReason) {
```
Comment (sven1103) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleSpreadsheetLayout.java:165]:
```suggestion
    private void displayErrorInstructions(String instructions) {
```

Comment (sven1103) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleSpreadsheetLayout.java:165]: or

Comment (sven1103) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleSpreadsheetLayout.java:165]:
```suggestion
    private void displayErrorInstructions(String reason) {
```

**Review by sven1103 — CHANGES_REQUESTED** [review]
Some minor change requests, and I will checkout and test the behavior in the meantime :)

**Comment by sven1103** [comment]
Behavior is as intended, validation errors are removed, after cells have been provided with the mandatory input.

This solution is way more intuitive, great job!

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java (hunk 1 above)
Comment (wow-such-code) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java:461]: The valid cells are recorded so the red highlighting can be removed when a user presses the registration button (and the cell was highlighted for missing data before). Unfortunately there is no possibility to remove it directly when inputting data, from what I can tell.

**Review by sven1103 — APPROVED** [review]

**Pull request merged**